src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs

```diff
@@ -1690,36 +1694,47 @@ public void UnmanagedConstraint_Compilation_Constructor()
+            );
+        }
+
+        [Fact]
+        public void UnmanagedConstraint_Compilation_AnotherType_Before1()
+        {
+            CreateCompilation("public class Test<T> where T : unmanaged, System.Enum { }").VerifyDiagnostics();
+
+            CreateCompilation("public class Test<T> where T : unmanaged, System.Delegate { }").VerifyDiagnostics(
+                // (1,19): error CS0455: Type parameter 'T' inherits conflicting constraints 'Delegate' and 'ValueType'
+                // public class Test<T> where T : unmanaged, System.Delegate { }
+                Diagnostic(ErrorCode.ERR_BaseConstraintConflict, "T").WithArguments("T", "System.Delegate", "System.ValueType").WithLocation(1, 19)
+                );
         }
 
         [Fact]
         public void UnmanagedConstraint_Compilation_AnotherType_After()
         {
             CreateCompilation("public class Test<T> where T : System.Exception, unmanaged { }").VerifyDiagnostics(
-                // (1,50): error CS8374: The 'unmanaged' constraint cannot be specified with other constraints.
+                // (1,50): error CS8374: The 'unmanaged' constraint must come before any other constraints
                 // public class Test<T> where T : System.Exception, unmanaged { }
-                Diagnostic(ErrorCode.ERR_UnmanagedConstraintMustBeAlone, "unmanaged").WithLocation(1, 50));
+                Diagnostic(ErrorCode.ERR_UnmanagedConstraintMustBeFirst, "unmanaged").WithLocation(1, 50));
         }
 
         [Fact]
         public void UnmanagedConstraint_Compilation_AnotherParameter_After()
         {
             CreateCompilation("public class Test<T, U> where T : U, unmanaged { }").VerifyDiagnostics(
-                // (1,38): error CS8374: The 'unmanaged' constraint cannot be specified with other constraints.
+                // (1,38): error CS8374: The 'unmanaged' constraint must come before any other constraints
                 // public class Test<T, U> where T : U, unmanaged { }
-                Diagnostic(ErrorCode.ERR_UnmanagedConstraintMustBeAlone, "unmanaged").WithLocation(1, 38));
+                Diagnostic(ErrorCode.ERR_UnmanagedConstraintMustBeFirst, "unmanaged").WithLocation(1, 38));
         }
 
         [Fact]
         public void UnmanagedConstraint_Compilation_AnotherParameter_Before()
         {
-            CreateCompilation("public class Test<T, U> where T : unmanaged, U { }").VerifyDiagnostics(
-                // (1,46): error CS8374: The 'unmanaged' constraint cannot be specified with other constraints.
-                // public class Test<T, U> where T : unmanaged, U { }
-                Diagnostic(ErrorCode.ERR_UnmanagedConstraintMustBeAlone, "U").WithLocation(1, 46));
+            CreateCompilation("public class Test<T, U> where T : unmanaged, U { }").VerifyDiagnostics();
+            CreateCompilation("public class Test<T, U> where U: class where T : unmanaged, U, System.IDisposable { }").VerifyDiagnostics();
```

Comment:
> CreateCompilation("public class Test<T, U> where U: class where T : unmanaged, U, System.IDisposable { }").VerifyDiagnostics(); [](start = 11, length = 128)

Should we get an error here given that the ```U``` has a class constraint?

Reply:
No. The behavior of `unmanaged` here is similar to `struct`. The `class` is not a transitive constraint - since U can be an object and T still can be anything.

---
In reply to: [173906250](https://github.com/dotnet/roslyn/pull/25397#discussion_r173906250) [](ancestors = 173906250)